man/man1/g2s.1
.TH g2s 1falcON "16 Feb 2007"

.SH name
\fBg2s\fP \- converts gadget to nemo snapshot file(s) using falcON

.SH synopsis
\fBgagdet2nemo\fP [parameter=value]

.SH requirement
requires NEMO version 3.0.13 or higher, see
http://www.astro.umd.edu/nemo.

.SH description
\fBg2s\fP tries to read gadget snapshots (gadget data
format 1) and write them into a single nemo snapshot file. If the
input has been different endianess than the local machine, the
data are converted on the fly.

.SH parameters

.TP 12
\fBin=\fP\fIfbase\fP
Gives the base file name for gadget snapshot data files. Default:
\fBrequired\fP.
.TP
\fBout=\fP\fIfile\fP
File name for writing data in nemo snapshot format. Default: \fBrequired\fP.
.TP
\fBnshot=\fP\fIval\fP
Number of snapshots to be converted. If \fBnshot\fP>1, we assume that
there are snapshots \fIfbase\fP\fBXXX\fP with \fBXXX\fP running from
\fBfirst\fP to \fBnshot\fP-\fBfirst\fP+1. Otherwisw (if
\fBnshot\fP==1), we assume the snapshot \fIfbase\fP to exists. Note
that each snapshot can be distributed over several files, e.g. the 4th
snapshot may be in 8 files "\fIfbase\fP004.0" to "\fIfbase\fP004.7".
Default: \fB1\fP.
.TP
\fBcopy=\fP\fIIOstring\fP
Specifies which data are copied to the nemo snapshot file (and hence are
assumed present in the gadget data files).

We assume that the gadget file contains the following data in this
order: positions, velocities, keys (ids), masses (if not specified via
the header), internal gas energies, gas densities, gas smoothing
lengths, potentials, and accelerations. We stop reading the gadget
data file once all data to be copied have been read (or the end of the
file has been reached). Note that the first five are guaranteed to be
always present. Default: \fBmxvU\fP.
.TP
\fBfirst=\fP\fIval\fP
If \fBnshot\fP>1, we assume the first snapshot to be converted to have index
\fIval\fP. So, if e.g.\ \fBfirst\fP=4 and \fBnshot\fP=5, we seek to 
read snapshots "\fIfbase\fP004" to "\fIfbase\fP008". Default: \fB0\fP.
.TP
\fBheader=\fP\fIval\fP
Size (in bytes) of the FORTRAN record header used when the gadget data
files have been written. This is usually 4, but on some 64bit machines it
may be 8. Using a wrong value is likely to result in a fatal error with
a completely misguiding error message. Default: \fB4\fP.

.SH limitations
Gadget is a massively parallel code and a single gadget snapshot distributed
over several files may contain more data than can be hold in memory on just
one node. Such snapshots cannot presently converted; the code is likely to
crash when trying to grab too much memory.

.SH see also
s2g(1falcON)
.PP

.fi
.SH files
.ta +3i
.nf
falcON/doc/user_guide.pdf               \fIUser Guide for falcON\fP.
.fi
.SH author
Walter Dehnen   walter.dehnen@astro.le.ac.uk
.SH update history of this man page
.nf
.ta +1.0i +2.0i
16-Feb-2007	first version of manual  WD
.fi




usr/dehnen/falcON/src/public/exe/g2s.cc
// -*- C++ -*-                                                                 |
//-----------------------------------------------------------------------------+
//                                                                             |
// g2s.cc                                                                      |
//                                                                             |
// Copyright (C) 2007,2008 Walter Dehnen                                       |
//                                                                             |
// This program is free software; you can redistribute it and/or modify        |
// it under the terms of the GNU General Public License as published by        |
// the Free Software Foundation; either version 2 of the License, or (at       |
// your option) any later version.                                             |
//                                                                             |
// This program is distributed in the hope that it will be useful, but         |
// WITHOUT ANY WARRANTY; without even the implied warranty of                  |
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU           |
// General Public License for more details.                                    |
//                                                                             |
// You should have received a copy of the GNU General Public License           |
// along with this program; if not, write to the Free Software                 |
// Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.                   |
//                                                                             |
//-----------------------------------------------------------------------------+
//                                                                             |
// Gadget to NEMO snapshot file converter                                      |
//                                                                             |
//-----------------------------------------------------------------------------+
//                                                                             |
// history:                                                                    |
//                                                                             |
// v 0.0    14/02/2007  WD created & seriously debugged                        |
// v 0.1    14/02/2007  WD added param header                                  |
// v 0.2    15/02/2007  WD can handle different endianess; added param first   |
// v 0.3    16/02/2007  WD checks on file size, only writes what it got        |
// v 1.0    16/02/2007  WD moved to public part of falcON                      |
// v 1.0.1  23/02/2007  WD renamed to g2s (previously gadget2nemo)             |
// v 1.0.2  10/09/2008  WD happy gcc 4.3.1                                     |
//-----------------------------------------------------------------------------+
#define falcON_VERSION   "1.0.2"
#define falcON_VERSION_D "10-sep-2008 Walter Dehnen                          "
//-----------------------------------------------------------------------------+
#include <main.h>                                  // main & NEMO stuff         
#include <fstream>                                 // C++ file I/O              
#include <iomanip>     
#include <string>
////////////////////////////////////////////////////////////////////////////////
const char*defv[] = {	
  "in=???\n           input base file (GADGET format)                    ",
  "out=???\n          output file (nemo snapshot format)                 ",
  "nshot=1\n          number of snapshots to read                        ",
  "copy=mxvU\n        data to copy (maximum: mxvkURHpa)                  ",
  "first=0\n          first snapshot to read                             ",
  "header=4\n         header size or unfio (4 or 8)                      ",
  falcON_DEFV, NULL};
const char*usage="GADGET to NEMO converter (better than \"gadget2nemo\")";
////////////////////////////////////////////////////////////////////////////////
void falcON::main() falcON_THROWING
{
  const char* fbase=getparam("in");
  unsigned nshot=getuparam("nshot");
  unsigned header=getuparam("header");
  nemo_out out(getparam("out"));
  fieldset copy=getioparam("copy") & fieldset("mxvkURHpa"), got;
  snapshot shot;
  if(nshot > 1) {
    for(unsigned i=0; i!=nshot; ++i) {
      char file[256];
      sprintf(file, "%s%03d", fbase, i+getuparam("first"));
      got = shot.read_gadget(file, copy, header);
      shot.write_nemo(out, copy & got);
    }
  } else {
    got = shot.read_gadget(fbase, copy, header);
    shot.write_nemo(out, copy & got);
  }
}
////////////////////////////////////////////////////////////////////////////////
